### PR TITLE
Take initial guess on max number of neighbors when building neighbor lists with ArborX

### PR DESCRIPTION
**Pull request by dalg24**
**Motivation:**  Pre-allocate memory and attempt to fill and count in a single pass.  Fallback to a second pass is storage was insufficient.

**Changes:**
* Added `buffer_size` trailing argument that defaults to zero (i.e. no buffer optimization) for `makeNeighborList` and `make2DNeighborList`
* Leveraged ArborX buffer optimization capability when building compressed neighbor lists
* Implemented buffer optimization for 2D neighbor lists

### CODE REVIEW

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 1]
```diff
@@ -152,10 +152,9 @@ struct NeighborDiscriminatorCallback2D_FirstPass
 {
     Counts counts;
     using tag = ArborX::Details::InlineCallbackTag;
-    template <typename Predicate, typename OutputFunctor>
+    template <typename Predicate>
     KOKKOS_FUNCTION void operator()( Predicate const &predicate,
-                                     int primitive_index,
-                                     OutputFunctor const & ) const
+                                     int primitive_index ) const
```
Comment (dalg24) [inline core/src/Cabana_Experimental_NeighborList.hpp:157]: I am taking advantage of the new `BVH::query()` overload recently added (within last month), which require updating ArborX.  I can revert if you think that is an issue.

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 2]
```diff
@@ -165,17 +164,39 @@ struct NeighborDiscriminatorCallback2D_FirstPass
...
+    template <typename Predicate>
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index ) const
+    {
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
+        {
+            if ( counts( predicate_index ) < (int)neighbors.extent( 1 ) )
+            {
+                neighbors( predicate_index, counts( predicate_index )++ ) =
```
Comment (sslattery) [inline core/src/Cabana_Experimental_NeighborList.hpp:183]: is this operation on `counts` atomic? does it need to be?

Comment (dalg24) [reply core/src/Cabana_Experimental_NeighborList.hpp:183]: See comment
https://github.com/ECP-copa/Cabana/blob/9cf488b7e8ec7dbda6616787e44c9b58390b6054/core/src/Cabana_Experimental_NeighborList.hpp#L211-L212

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 3]
```diff
@@ -165,17 +164,39 @@ struct NeighborDiscriminatorCallback2D_FirstPass
...
+    Neighbors neighbors;
+    using tag = ArborX::Details::InlineCallbackTag;
+    template <typename Predicate>
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index ) const
+    {
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
+        {
+            if ( counts( predicate_index ) < (int)neighbors.extent( 1 ) )
```
Comment (aprokop) [inline core/src/Cabana_Experimental_NeighborList.hpp:181]: From the performance perspective, you don't want to call extent all the time here. Why not store it as a member?

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 4]
```diff
@@ -165,17 +164,39 @@ struct NeighborDiscriminatorCallback2D_FirstPass
     }
 };
 
+// Preallocate and attempt fill in the first pass
+template <typename Counts, typename Neighbors, typename Tag>
+struct NeighborDiscriminatorCallback2D_FirstPass_BufferOptimization
+{
+    Counts counts;
+    Neighbors neighbors;
+    using tag = ArborX::Details::InlineCallbackTag;
```
Comment (aprokop) [inline core/src/Cabana_Experimental_NeighborList.hpp:173]: I thought we got rid of these?

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 5]
```diff
@@ -238,40 +260,60 @@ template <typename DeviceType, typename Slice, typename Tag>
...
+    {
+        neighbors = Kokkos::View<int **, DeviceType>( "neighbors", n_queries,
+                                                      buffer_size );
+        bvh.query(
+            space, predicates,
+            Impl::NeighborDiscriminatorCallback2D_FirstPass_BufferOptimization<
+                decltype( counts ), decltype( neighbors ), Tag>{counts,
+                                                                neighbors} );
+    }
+    else
```
Comment (aprokop) [inline core/src/Cabana_Experimental_NeighborList.hpp:291]: So you don't allow negative buffer here? Maybe at least check?

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 6]
```diff
@@ -238,40 +260,60 @@ template <typename DeviceType, typename Slice, typename Tag>
...
+    Kokkos::View<int **, DeviceType> neighbors;
     Kokkos::View<int *, DeviceType> counts( "counts", n_queries );
-    bvh.query(
-        space, predicates,
-        Impl::NeighborDiscriminatorCallback2D_FirstPass<decltype( counts ),
-                                                        Tag>{counts},
-        indices, offset );
+    if ( buffer_size > 0 )
+    {
+        neighbors = Kokkos::View<int **, DeviceType>( "neighbors", n_queries,
```
Comment (aprokop) [inline core/src/Cabana_Experimental_NeighborList.hpp:283]: Should use `WithoutInitializing`

File: core/src/Cabana_Experimental_NeighborList.hpp (hunk 4 above)
Comment (dalg24) [reply core/src/Cabana_Experimental_NeighborList.hpp:173]: It is being ignored.  But yes can safely be removed.

File: core/src/Cabana_Experimental_NeighborList.hpp (hunk 6 above)
Comment (dalg24) [reply core/src/Cabana_Experimental_NeighborList.hpp:283]: Thx

File: core/src/Cabana_Experimental_NeighborList.hpp (hunk 5 above)
Comment (dalg24) [reply core/src/Cabana_Experimental_NeighborList.hpp:291]: Sure will assert it is > 0

File: core/src/Cabana_Experimental_NeighborList.hpp (hunk 3 above)
Comment (dalg24) [reply core/src/Cabana_Experimental_NeighborList.hpp:181]: I don't think it matters although I have no more evidence than you do.  I am not changing.

File: core/src/Cabana_Experimental_NeighborList.hpp (hunk 1 above)
Comment (streeve) [reply core/src/Cabana_Experimental_NeighborList.hpp:155]: I have no issue with that

File: core/src/Cabana_Experimental_NeighborList.hpp [hunk 7]
```diff
@@ -238,40 +258,63 @@ template <typename DeviceType, typename Slice, typename Tag>
...
+    {
+        // NOTE We do not bother shrinking to eliminate the excess allocation.
+        // NOTE If buffer_size is 0, neighbors is default constructed.  This is
+        // fine with the current design/implementation of NeighborList access
+        // traits.
+        return Dense<MemorySpace, Tag>{counts, neighbors, first, bvh.size()};
+    }
+
+    neighbors = Kokkos::View<int **, DeviceType>(
         Kokkos::view_alloc( "neighbors", Kokkos::WithoutInitializing ),
```
Comment (streeve) [inline core/src/Cabana_Experimental_NeighborList.hpp:311]: Should we start using `view_alloc` throughout?

Comment (dalg24) [reply core/src/Cabana_Experimental_NeighborList.hpp:311]: Up to you.  At this point I am not sure we will actually ever go further than kokkos/kokkos#3264 and actually pull the trigger on deprecating `ViewAllocateWithoutInitializing`.

**Comment by aprokop** [comment]
Merge this?

**Pull request merged**